src/cc/api/BPFTable.cc

```diff
@@ -180,6 +180,66 @@ StatusTuple BPFTable::clear_table_non_atomic() {
   return StatusTuple(0);
 }
 
+StatusTuple BPFTable::get_table_offline(
+  std::vector<std::pair<std::string, std::string>> &res) {
+  StatusTuple r(0);
+
+  auto key = std::unique_ptr<void, decltype(::free)*>(::malloc(desc.key_size),
+                                                      ::free);
+  auto value = std::unique_ptr<void, decltype(::free)*>(::malloc(desc.leaf_size),
+                                                      ::free);
+  std::string key_str;
+  std::string value_str;
+
+  if (desc.type == BPF_MAP_TYPE_ARRAY ||
+      desc.type == BPF_MAP_TYPE_PROG_ARRAY ||
+      desc.type == BPF_MAP_TYPE_PERF_EVENT_ARRAY ||
+      desc.type == BPF_MAP_TYPE_PERCPU_ARRAY ||
+      desc.type == BPF_MAP_TYPE_CGROUP_ARRAY ||
+      desc.type == BPF_MAP_TYPE_STACK_TRACE ||
+      desc.type == BPF_MAP_TYPE_ARRAY_OF_MAPS ||
+      desc.type == BPF_MAP_TYPE_DEVMAP ||
+      desc.type == BPF_MAP_TYPE_CPUMAP ||
+      desc.type == BPF_MAP_TYPE_REUSEPORT_SOCKARRAY) {
```

Comment:
Hi @yonghong-song , 

Using the `get_next_key()` approach would be the same as in the array based elements this function just returns `index + 1`: https://git.kernel.org/pub/scm/linux/kernel/git/bpf/bpf-next.git/tree/kernel/bpf/arraymap.c#n232.

I think a possible solution is to skip the element if lookup fails, instead of breaking the loop as it currently is.

What do you think it is better, to return a vector of the size of the array with empty elements `std::pair<std::string, std::string>("", "")`, or just to return the valid elements in the array?

Reply:
Thanks for pointing out. You are right. The only exception is `BPF_MAP_TYPE_STACK_TRACE` which you get the real next id. The `BPF_MAP_TYPE_STACK_TRACE` is considered as a hash table is not an array.

For all other array types, you can check the return value/error, if the error is ENOENT, you can move on to the next one. Otherwise, an error should be returned. 

I think we should return only valid elements in the array. There is no need for extra storage for non-existing keys.

Reply:
I agree, done!